conbench/entities/history.py

```diff
@@ -85,31 +89,170 @@ def get_history(case_id, context_id, hardware_hash, repo) -> List[tuple]:
+        .over(order_by=history.c.timestamp)
+        .label("rolling_sd_including_this_commit"),
+    ).subquery()
+
+    # Add same column, excluding the current commit
+    history = Session.query(
+        history,
+        func.stddev(history.c.residual)
+        .over(
+            order_by=s.extract("epoch", history.c.timestamp),
+            range_=(None, -1),
+        )
+        .label("rolling_sd_excluding_this_commit"),
+    )
+
+    return history.all()
+
+
+def _calc_z_score(
+    data_point: float, unit: str, dist_mean: Optional[float], dist_sd: Optional[float]
+) -> Optional[float]:
+    """Calculate the z-score of a data point compared to a distribution."""
+    if dist_mean is not None and dist_sd is not None and dist_sd != 0:
+        z_score = (data_point - dist_mean) / dist_sd
+    else:
+        z_score = None
+
+    if z_score and _less_is_better(unit):
+        z_score = z_score * -1
+
+    return z_score
+
+
+def append_z_scores(benchmark_results: List[BenchmarkResult]):
+    """Populate the "z_score" attribute of each given BenchmarkResult, comparing it to
+    the distribution of BenchmarkResults that share the same hardware/case/context/repo,
+    ending with the most recent BenchmarkResult in the given BenchmarkResult's commit
+    ancestry on the default branch.
+
+    We choose a commit from the default branch as the end commit, because we probably
+    don't want "bad" commits early on in a PR to set artificially low z-scores later in
+    the PR.
+
+    Should not raise anything. If we can't find a z-score for some reason, the z_score
+    attribute will be None on that BenchmarkResult.
+    """
+    # {(case, context, hardware, repo, commit): (mean, sd)}
+    cached_stats: Dict[
+        Tuple[str, str, str, str, str],
+        Tuple[Optional[float], Optional[float]],
+    ] = {}
+
+    for benchmark_result in benchmark_results:
+        if benchmark_result.error or benchmark_result.mean is None:
+            benchmark_result.z_score = None
+            continue
+
+        baseline_run = benchmark_result.run.get_baseline_run(
+            on_default_branch=True,
+            case_id=benchmark_result.case_id,
+            context_id=benchmark_result.context_id,
+        )
+        if not baseline_run:
+            log.debug(
+                f"BenchmarkResult {benchmark_result.id} has no baseline run; "
+                "setting z_score = None"
+            )
+            benchmark_result.z_score = None
+            continue
+
+        case = benchmark_result.case_id
+        context = benchmark_result.context_id
+        hardware = benchmark_result.run.hardware.hash
+        repo = benchmark_result.run.commit.repository
+        baseline_sha = baseline_run.commit.sha
+
+        if (case, context, hardware, repo, baseline_sha) not in cached_stats:
+            # Fill in the cache with all the stats for this history
+            history = get_history(case, context, hardware, repo)
```

Comment:
Checking my understanding: By this comment, I'm assuming that this in practice runs once and then uses cached results, even though this gets called for every `BenchmarkResult`? Otherwise I was worried about this function repeatedly doing the same work, but if there's caching it should be fast.

Reply:
Yes, see this discussion: https://github.com/conbench/conbench/pull/636#discussion_r1090929658